imap_processing/idex/idex_l1b.py

```diff
@@ -0,0 +1,175 @@
+    # Copy arrays from the l1a_dataset that do not need l1b processing
+    for var in vars_to_copy:
+        l1b_dataset[var] = l1a_dataset[var].copy()
+
+    l1b_dataset["epoch"] = epoch_da
+
+    # TODO: Add TriggerMode and TriggerLevel attr
+    # TODO: Spice data?
+
+    logger.info("IDEX L1B science data processing completed.")
+
+    return l1b_dataset
+
+
+def unpack_instrument_settings(
+    l1a_dataset: xr.Dataset,
+    var_information_df: pd.DataFrame,
+    idex_attrs: Optional[ImapCdfAttributes] = None,
+) -> dict[str, xr.DataArray]:
+    """
+    Unpack raw telemetry data from the l1a dataset into individual variables.
+
+    Parameters
+    ----------
+    l1a_dataset : xarray.Dataset
+        IDEX L1a dataset containing the 6 waveform arrays.
+    var_information_df : pd.DataFrame
+        Pandas data frame that contains information about each variable
+        (e.g., bit-size, starting bit, and padding). This is used to unpack raw
+        telemetry data from the input dataset (`l1a_dataset`).
+    idex_attrs : ImapCdfAttributes
+        CDF attribute manager object.
+
+    Returns
+    -------
+    telemetry_das : dict
+        A dictionary where the keys are the instrument setting array names and the
+        values are the unpacked xr.DataArrays.
+    """
+    telemetry_das = {}
+
+    for _, row in var_information_df.iterrows():
+        var_name = row["mnemonic"]
+
+        # Create binary mask of the size of the variable in bits
+        mask = (1 << row["unsigned_nbits"]) - 1
+        # Determine the number of bits to shift
+        shift = row["starting_bit"] - row["nbits_padding_before"]
+        # Get the unpacked value by shifting the data to align the desired bits with
+        # the least significant bits and applying the mask to isolate the target bits
+        unpacked_val = (l1a_dataset[row["packetName"]].data >> shift) & mask
+
+        attrs = idex_attrs.get_variable_attributes(var_name) if idex_attrs else None
+
+        telemetry_das[var_name] = xr.DataArray(
+            name=var_name,
+            data=unpacked_val,
+            dims=("epoch"),
+            attrs=attrs,
+        )
+
+    return telemetry_das
+
+
+def convert_waveforms(
+    l1a_dataset: xr.Dataset, idex_attrs: ImapCdfAttributes
+) -> dict[str, xr.DataArray]:
+    """
+    Apply transformation from raw dn to picocoulombs (pC) for each of the six waveforms.
+
+    Parameters
+    ----------
+    l1a_dataset : xarray.Dataset
+        IDEX L1a dataset containing the six waveform arrays.
+    idex_attrs : ImapCdfAttributes
+        CDF attribute manager object.
+
+    Returns
+    -------
+    waveforms_converted : dict
+        A dictionary where the keys are the waveform array names and the values are
+        xr.DataArrays representing the waveforms transformed into picocoulombs.
+    """
+    waveforms_pc = {}
+
+    for var in ConvFactors:
+        waveforms_pc[var.name] = l1a_dataset[var.name] * var.value
```

Comment:
I *think* that numpy will convert the dtype to a `float` in this statement. Is that the desired dtype for the output CDF?

Reply:
Yes the output should now be a float64 (matching the l1b validation file)